### ISSUE TITLE
Strange NullPointerException when printing converted text

### ISSUE THREAD

**Issue by JustusAdam** [typelead/eta#490]
So i have this really obscure error, where printing a value which was converted to Haskell from Java/Clojure results in a `NullPointerException` but *only* if the value was not previously evaluated.

Basically this works

```hs
main = do
    !r <- Clojure.read "a"
    let converted = fromNative r
    hPutStrLn stderr $ fromJava $ toString r
    let Sym s = converted
    hPutStrLn stderr $ show $ s == Symbol Nothing "a" -- equality check forces evaluation
    hPutStrLn stderr $ show s
```

And produces the output you'd expect

```
a
True
Symbol {namespace = Nothing, name = "a"}
```

And this, which only elides the equality check raises a `NullPointerException`

```hs
main = do
    !r <- Clojure.read "a"
    let converted = fromNative r
    hPutStrLn stderr $ fromJava $ toString r
    let Sym s = converted
    hPutStrLn stderr $ show s
```

```
a
Symbol {namespace = Nothing, name = "a"Exception in thread "main" java.lang.NullPointerException
        at base.ghc.io.handle.Text$hPutStr3.enter(Unknown Source)
        at base.ghc.io.handle.Text$hPutStr2.enter(Unknown Source)
        at main.Main$main1.enter(Unknown Source)
        at main.Main$DZCmain.enter(Unknown Source)
        at eta.runtime.apply.Function.applyV(Function.java:16)
        at eta.runtime.stg.Closures$EvalLazyIO.enter(Closures.java:100)
        at eta.runtime.stg.Capability.schedule(Capability.java:153)
        at eta.runtime.stg.Capability.scheduleClosure(Capability.java:100)
        at eta.runtime.Runtime.evalLazyIO(Runtime.java:189)
        at eta.runtime.Runtime.main(Runtime.java:182)
        at eta.main.main(Unknown Source)
```

## Some additional information

```
data Symbol = Symbol 
    { namespace :: Maybe T.Text 
    , name :: T.Text
    } deriving Eq
```

The implementation of `fromNative` and the associated helper functions can be found [here](https://github.com/ohua-dev/ohua-jvm-integration/blob/master/src/eta/Ohua/Compat/JVM/Marshal.hs) and the raw clojure interop [here](https://github.com/ohua-dev/ohua-jvm-integration/blob/master/src/eta/Clojure.hs).

**Comment by rahulmutt** [comment]
Thanks for the report! I'll try reproducing this locally - I just have to put this code in `ohua-jvm-integration` and run it correct?

**Comment by JustusAdam** [comment]
Give me a second I'll push some fixes from today. I also got an even cleaner version of the error using `deepseq` to evaluate.

**Comment by rahulmutt** [comment]
Do any of the Clojure APIs return `null`?

**Comment by rahulmutt** [comment]
Btw you can do `etlas run --trace > TRACE 2>&1` (redirected because the output can be huge) to get a full on stack trace and track down where the `null` originated.

**Comment by rahulmutt** [comment]
Hmm looks like the trace is useless in this case, since it seems the Clojure runtime is interacting badly with slf4j and that causes a StackOverflowError. I'll see if I can track it down with VisualVM.

**Comment by JustusAdam** [comment]
They shouldn't return `nil` I mean they do occasionally but in this case all operations should be safe.

**Comment by rahulmutt** [comment]
I've been taking a look at this. The existing debugging mechanisms (`etlas run --trace` and VisualVM) offered no insight into where the null is being created so I had to create a simple jdb scripting DSL (written in Haskell of course) that essentially automates a jdb workflow and creates a precision trace to figure out what's going on. I just got that working yesterday, so now it's a matter of going through the traces that were produced.

**Comment by JustusAdam** [comment]
Cool. Thx for the update.

**Comment by rahulmutt** [comment]
Going through the trace isn't revealing any answers. All I know is that the failure is happening only when you use the `cljName` function *inside* of `Symbol` which I found weird. Using it alone works fine. The trace reveals that the exception occurs when running `Text.unpack` - so at the last step of displaying an evaluated Clojure value. Interesting part is when trying to trace the null, I'm unable to find the source - somewhere in the middle of backtracking, there's a missing link.

I've reduced it to this:
```haskell
main = do
    r <- Clojure.read "a"
    print $ Symbol Nothing (cljName r)
```
I found this after trying out different inputs for `read` and finding that only the `Sym` case yields problems.

Have you been encountering more of this in different ways or is this just a very obscure bug that can be worked around?
